# AEACNA - Matemática
Atividade Especial para Acompanhamento Continuo do Nivel de Aprendizagem

## Instruções

1. Enviar para o email filipe.furlan@letscode.com.br até **2021/08/17 as 16:00** o notebook preenchido (avalicao_matematica_firstname_lastname.ipynb) com a resolução dos exercicios abaixo.


2. Atividade individual com consulta ao material das aulas e livros/links de referência.


# Questão 01 - 2,0

Carregue os dados do dataset [`car_details`](https://s3-sa-east-1.amazonaws.com/lcpi/4fbc4943-7af3-4c42-9a21-4d9eb121a28d.csv) e, em seguida:

a. Classifique as variáveis do dataset como quantitativas (contínua/discreta) ou qualitativas (ordinal/nominal);  
b. Obtenha a média, a mediana e a moda das variáveis quantitativas e, em seguida, classifique cada uma dessas variáveis quanto à sua simetria.  
c. Calcule os percentis do preço de venda dos carros. Em seguida, indique o percentual dos carros que podem ser adquiridos por um preço de até R\\$ 120.000,00.  
d. Crie uma tabela de distribuição de frequências para a informação do preço de venda dos carros. Para isso, defina o número de classes, utilizando a Regra de Sturges.

In [1]:
import pandas as pd
import numpy as np
from scipy.stats import norm
from scipy.stats import t as t_student

In [2]:
df = pd.read_csv('car_details.csv')
df.head()

name  year  selling_price  km_driven    fuel  \
0             Maruti 800 AC  2007          60000      70000  Petrol   
1  Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
2      Hyundai Verna 1.6 SX  2012         600000     100000  Diesel   
3    Datsun RediGO T Option  2017         250000      46000  Petrol   
4     Honda Amaze VX i-DTEC  2014         450000     141000  Diesel   

  seller_type transmission         owner  
0  Individual       Manual   First Owner  
1  Individual       Manual   First Owner  
2  Individual       Manual   First Owner  
3  Individual       Manual   First Owner  
4  Individual       Manual  Second Owner

## Variáveis quantitativas

> Contínuas:
selling_price, km_driven

## Variáveis qualitativas

> Nominais:
name, fuel, seller_type, transmission, owner

> Ordinais:
year

## Média, mediana e moda

In [3]:
mediaPreco = df['selling_price'].mean()
mediaKM = df['km_driven'].mean()
print(f'Média de Preços: {mediaPreco}')
print(f'Média de KMs: {mediaKM}')

medianaPreco = df['selling_price'].median()
medianaKM = df['km_driven'].median()
print(f'Mediana de Preços: {medianaPreco}')
print(f'Mediana de KMs: {medianaKM}')

modaPreco = df['selling_price'].mode()
modaKM = df['km_driven'].mode()
print(f'Moda de Preços: {modaPreco}')
print(f'Moda de KMs: {modaKM}')

Média de Preços: 504127.3117511521
Média de KMs: 66215.77741935484
Mediana de Preços: 350000.0
Mediana de KMs: 60000.0
Moda de Preços: 0    300000
dtype: int64
Moda de KMs: 0    70000
dtype: int64


## Porcentual de carros vendidos até R$ 120.000,00

In [4]:
df2c = df
df2c['percentis'] = df2c['selling_price']/df2c['selling_price'].sum()
df2c = df2c[df2c['selling_price'] <= 120000]
percentual = df2c['percentis'].sum()
print(f'Porcentual de carros vendidos até R$ 120.000,00: {percentual*100:.2f}%')

Porcentual de carros vendidos até R$ 120.000,00: 2.01%


## Tabela de distribuição de frequências

In [5]:
n = df.shape[0]

k = round(1 + 10 / 3 * np.log10(n))

dfSturges = df
dfSturges['selling_price'] = dfSturges['selling_price'].astype(float)
dfSturges['classificacao'] = pd.cut(dfSturges['selling_price'], bins = k)
dfSturges['selling_price'].replace(to_replace = 0, value = np.nan, inplace=True)
dfSturges = dfSturges.dropna()

fi = dfSturges['classificacao'].value_counts()
fr = dfSturges['classificacao'].value_counts() / dfSturges.shape[0]
fp = fr * 100
#tabela
dist_freq_renda = pd.DataFrame({'Frequência Absoluta (fi)': fi, 'Frequência Percentual (fp)': fp})
dist_freq_renda['Frequência Acumulada (fa)']  = dist_freq_renda['Frequência Percentual (fp)'].cumsum()
dist_freq_renda

Frequência Absoluta (fi)  \
(11120.0, 703076.923]                           3571   
(703076.923, 1386153.846]                        578   
(1386153.846, 2069230.769]                        96   
(2069230.769, 2752307.692]                        39   
(3435384.615, 4118461.538]                        18   
(2752307.692, 3435384.615]                        15   
(4801538.462, 5484615.385]                        11   
(4118461.538, 4801538.462]                         9   
(5484615.385, 6167692.308]                         1   
(7533846.154, 8216923.077]                         1   
(8216923.077, 8900000.0]                           1   
(6167692.308, 6850769.231]                         0   
(6850769.231, 7533846.154]                         0   

                            Frequência Percentual (fp)  \
(11120.0, 703076.923]                        82.281106   
(703076.923, 1386153.846]                    13.317972   
(1386153.846, 2069230.769]                    2.211982   
(2069230.769, 2752307.692]                    0.898618   
(3435384.615, 4118461.538]                    0.414747   
(2752307.692, 3435384.615]                    0.345622   
(4801538.462, 5484615.385]                    0.253456   
(4118461.538, 4801538.462]                    0.207373   
(5484615.385, 6167692.308]                    0.023041   
(7533846.154, 8216923.077]                    0.023041   
(8216923.077, 8900000.0]                      0.023041   
(6167692.308, 6850769.231]                    0.000000   
(6850769.231, 7533846.154]                    0.000000   

                            Frequência Acumulada (fa)  
(11120.0, 703076.923]                       82.281106  
(703076.923, 1386153.846]                   95.599078  
(1386153.846, 2069230.769]                  97.811060  
(2069230.769, 2752307.692]                  98.709677  
(3435384.615, 4118461.538]                  99.124424  
(2752307.692, 3435384.615]                  99.470046  
(4801538.462, 5484615.385]                  99.723502  
(4118461.538, 4801538.462]                  99.930876  
(5484615.385, 6167692.308]                  99.953917  
(7533846.154, 8216923.077]                  99.976959  
(8216923.077, 8900000.0]                   100.000000  
(6167692.308, 6850769.231]                 100.000000  
(6850769.231, 7533846.154]                 100.000000

# Questão 02 - 2,0



De 50.000 válvulas fabricadas por uma companhia retira-se uma amostra de 400 válvulas, e obtém a vida média de 800 horas e o desvio padrão de 100 horas.

a. Qual o intervalo de confiança de 99% para a vida média da população? Explique com suas palavras o que esse intervalo de confiança significa.  
b. Que tamanho deve ter a amostra para que o erro máximo seja de 7,84 horas, com uma confiança de 95%?  

## Intervalo de confiança

In [6]:
media_amostral = 800
sigma = 100
n = 400
confianca = 0.99
intervalo = norm.interval(alpha = confianca, loc=media_amostral, scale=sigma / np.sqrt(n))
print(f'Intervalo de confiança de 99%: {intervalo}')

Intervalo de confiança de 99%: (787.1208534822555, 812.8791465177445)


> Supondo uma distribuição normal, com a média de 800 horas e desvio padrão de 100 horas da amostra, há uma probabilidade de 99% de que uma válvula fábricada pela companhia tenha a vida média entre 787,12 e 812,88 horas.

## Tamanho amostral

In [7]:
erro_maximo = 7.84

z_a2 = 2.57 # da tabela

n = round((z_a2 * sigma / erro_maximo)**2)

print(f'Tamanho da amostra: {n}')

Tamanho da amostra: 1075


# Questão 03 - 2,0


Na indústria cerâmica, avalia-se sistematicamente a resistência de amostras de massas cerâmicas, após o processo de queima. Dessas avaliações, sabe-se que certo tipo de massa tem resistência mecânica aproximadamente normal, com **média populacional de 53 MPa** e **variância populacional de 16 MPa²**. Após a troca de alguns fornecedores de matérias-primas, deseja-se **verificar se houve alteração na qualidade** (se a média deixou de ser 53 MPa). Uma **amostra de 15** corpos de prova de massa cerâmica acusou **média igual a 50 MPa**. Qual é a conclusão ao nível de significância de 5%?

In [8]:
u = 53
sigma = 16**0.5
# H0: média = 53
# H1: média != 53

n = 15
graus_de_liberdade = n - 1
x_barra = 50

significancia = 0.05/2
confianca = 1 - significancia

probabilidade = confianca + significancia/2

t_alpha = t_student.ppf(probabilidade, graus_de_liberdade)
t_alpha

2.5095694114796623

In [9]:
t = (x_barra - u) / (sigma / np.sqrt(n))
t

-2.904737509655563

In [10]:
if (t <= -1*t_alpha) or (t >= t_alpha):
    print('Rejeitar H0')
else:
    print('Aceitar H0')

Rejeitar H0


> Conclusão:
De acordo com a amostra obtida, a média de 53 MPa não se manteve.

# Questão 04 - 2,0

Você é um pesquisador que estuda o mercado de trabalho brasileiro e resolve estudar as diferenças salariais dos trabalhadores dos estados do Rio de Janeiro e de São Paulo. Durante sua pesquisa você verifica que, aparentemente, os rendimentos dos trabalhadores no estado do Rio de Janeiro são mais baixos que os rendimentos dos trabalhadores no estado de São Paulo. Para confirmar esta conclusão realize um teste de hipótese de comparação de médias em cima de duas amostras de trabalhadores dos dois estados. Siga as seguintes etapas:

- Selecione duas amostras de **500 trabalhadores** para cada um dos dois estados (variável UF) para realizar o teste. Utilize o **parâmetro random_state=101**.
- Considere o **nível de significância de 5%**.
- Teste a hipótese de que a **renda média dos trabalhadores do Rio de Janeiro é menor que a renda média dos trabalhadores de São Paulo**.

> [Link para o dataset](https://s3-sa-east-1.amazonaws.com/lcpi/360d2eaa-970a-4204-8004-53bb42af101a.csv).

> **OBS.:** Os dados da dataset referentes à UF foram convertidos (de categóricos para numéricos), portanto, utilize a tabela abaixo para identificar SP e RJ:

|Código|Descrição|
|---|---|
|11|Rondônia|
|12|Acre|
|13|Amazonas|
|14|Roraima|
|15|Pará|
|16|Amapá|
|17|Tocantins|
|21|Maranhão|
|22|Piauí|
|23|Ceará|
|24|Rio Grande do Norte|
|25|Paraíba|
|26|Pernambuco|
|27|Alagoas|
|28|Sergipe|
|29|Bahia|
|31|Minas Gerais|
|32|Espírito Santo|
|33|Rio de Janeiro|
|35|São Paulo|
|41|Paraná|
|42|Santa Catarina|
|43|Rio Grande do Sul|
|50|Mato Grosso do Sul|
|51|Mato Grosso|
|52|Goiás|
|53|Distrito Federal|

In [11]:
df = pd.read_csv('datasets/pnad_2015.csv')
df.head()

UF  Sexo  Idade  Cor  Anos de Estudo  Renda    Altura
0  11     0     23    8              12    800  1.603808
1  11     1     23    2              12   1150  1.739790
2  11     1     35    8              15    880  1.760444
3  11     0     46    2               6   3500  1.783158
4  11     1     47    8               9    150  1.690631

In [12]:
sp = df[df['UF'] == 35].sample(n=500, random_state=101)
sp = sp['Renda']
rj = df[df['UF'] == 33].sample(n=500, random_state=101)
rj = rj['Renda']

In [13]:
# H0: renda média RJ >= renda média SP
# H1: renda média RJ < renda média SP

significancia = 0.05
confianca = 1 - significancia
probabilidade = confianca

n = sp.size + rj.size

grau_liberdade = n - 2

u_sp = sp.mean()
u_rj = rj.mean()

sigma_sp = sp.std()
sigma_rj = rj.std()

z = (u_rj - u_sp) / (sigma_rj**2/rj.size + sigma_sp**2/sp.size)**0.5

z_alpha = norm.ppf(probabilidade)

print(z, z_alpha)

if (z <= z_alpha*-1):
    print('Rejeitar H0')
else:
    print('Aceitar H0')

-2.255318273608561 1.6448536269514722
Rejeitar H0


> A renda média no Rio de Janeiro é menor do que a renda média em São Paulo

# Questão 05 - 2,0

Após a finalização de um curso técnico, um estudante pediu a seu professor uma carta de recomendação para o primeiro emprego. Uma pesquisa de mercado aponta que há 80% de chances de obter o trabalho se receber uma recomendação excelente (E), 40% de chances se a recomendação for boa (B) e apenas 10% de chances se for fraca (F). 

Conversando com ex-alunos, o estudante estimou que as chances de receber uma recomendação excelente (E), boa (B) e fraca (F) são, respectivamente, 70%, 20% e 10%.

Calcule a probabilidade:

a. de que ele vá receber uma oferta de trabalho.  
b. de a recomendação ter sido fraca, boa ou excelente dado que ele recebeu a oferta de trabalho.  
c. de a recomendação ter sido fraca, boa ou excelente dado que ele **não** recebeu a oferta de trabalho.  
d. As probabilidades de 70%, 20% e 10% respectivamente estimadas para a recomendação ser excelente (E), boa (B) ou fraca (F) foram alteradas nos dois itens acima, para mais ou para menos em função dos resultados que elas produziram. Como você explica a lógica dessas modificações?

In [14]:
tabela = pd.DataFrame(data=[['80%', '40%', '10%'], 
                            ['20%', '60%', '90%']], index=['Contratado', 'Não contratado'], columns=['Excelente (E),', 'Boa (B)', 'Fraca (F)'])
tabela

Excelente (E), Boa (B) Fraca (F)
Contratado                80%     40%       10%
Não contratado            20%     60%       90%

## a) 
Probabilidade de receber uma oferta de trabalho

In [15]:
# c = contratado
c_E = 0.8
c_B = 0.4
c_F = 0.1

# nc = não contratado
nc_E = 1 - c_E
nc_B = 1 - c_B
nc_F = 1 - c_F

# Ex-alunos
c_ex_E = 0.7
c_ex_B = 0.2
c_ex_F = 0.1
nc_ex_E = 1 - c_ex_E
nc_ex_B = 1 - c_ex_B
nc_ex_F = 1 - c_ex_F

P_c_E = c_E * c_ex_E
P_c_B = c_B * c_ex_B
P_c_F = c_F * c_ex_F

probabilidade_contratado = P_c_E + P_c_B + P_c_F
print(f'Probabilidade do aluno ser contratado: {probabilidade_contratado*100:.1f}%')

Probabilidade do aluno ser contratado: 65.0%


## b) 

Probabilidade de a recomendação ter sido fraca, boa ou excelente dado que ele recebeu a oferta de trabalho

In [16]:
probE = P_c_E/probabilidade_contratado
probB = P_c_B/probabilidade_contratado
probF = P_c_F/probabilidade_contratado

print('Dado que o aluno foi contratado:')
print(f'Probabilidade de recomendação excelente: {probE*100:.1f}%')
print(f'Probabilidade de recomendação boa: {probB*100:.1f}%')
print(f'Probabilidade de recomendação fraca: {probF*100:.1f}%')

Dado que o aluno foi contratado:
Probabilidade de recomendação excelente: 86.2%
Probabilidade de recomendação boa: 12.3%
Probabilidade de recomendação fraca: 1.5%


## c) 
Probabilidade de a recomendação ter sido fraca, boa ou excelente dado que ele NÃO recebeu a oferta de trabalho

In [17]:
P_nc_E = nc_E * nc_ex_E
P_nc_B = nc_B * nc_ex_B
P_nc_F = nc_F * nc_ex_F

probabilidade_nao_contratado = P_nc_E + P_nc_B + P_nc_F

probnE = P_nc_E/probabilidade_nao_contratado
probnB = P_nc_B/probabilidade_nao_contratado
probnF = P_nc_F/probabilidade_nao_contratado

print('Dado que o aluno não foi contratado:')
print(f'Probabilidade de recomendação excelente: {probnE*100:.1f}%')
print(f'Probabilidade de recomendação boa: {probnB*100:.1f}%')
print(f'Probabilidade de recomendação fraca: {probnF*100:.1f}%')

Dado que o aluno não foi contratado:
Probabilidade de recomendação excelente: 4.4%
Probabilidade de recomendação boa: 35.6%
Probabilidade de recomendação fraca: 60.0%


## d)
As probabilidades de 70%, 20% e 10% respectivamente estimadas para a recomendação ser excelente (E), boa (B) ou fraca (F) foram alteradas nos dois itens acima, para mais ou para menos em função dos resultados que elas produziram. Como você explica a lógica dessas modificações?

In [18]:
P_c_E + P_c_B + P_c_F

print(f'Probabilidade de ser contratado com uma carta excelente: {P_c_E*100:.1f}% (original: 70%)')
print(f'Probabilidade de ser contratado com uma carta boa: {P_c_B*100:.1f}% (original: 20%)')
print(f'Probabilidade de ser contratado com uma carta fraca: {P_c_F*100:.1f}% (original: 10%)')

Probabilidade de ser contratado com uma carta excelente: 56.0% (original: 70%)
Probabilidade de ser contratado com uma carta boa: 8.0% (original: 20%)
Probabilidade de ser contratado com uma carta fraca: 1.0% (original: 10%)


> No item "a" calculamos a probabilidade do aluno ser contratado, 65% de chance dado que ele recebeu uma carta de recomendação. À partir do momento que sabemos que essa é a chance dele ser contratado, e que ele foi contratado, a chance dele ter sido contratado por conta de uma carta de recomendação em específico corresponde à probabilidade dele ser contratado por conta daquela carta de recomendação, dividido pela chance dele ser contratado (por pertencer ao "universo" probabilístico do aluno ser contratado, e por esse universo ter ocorrido).